### PR TITLE
[NEW] First Version of OLAP Reporting

### PR DESCRIPTION
**Pull request by crwr45**
Adding basic OLAP reporting framework with support for:
 - Defining report records
 - Error reporting through project tracker
 - Synchronous or Asynchronous updating of reports
 - Custom report viewers
 - (Partial) Exporting from custom report viewers.

Signed-off-by: Charlie Wheeler-Robinson <crwr@pigotts.org.uk>

### CODE REVIEW

File: addons/olap/models/report_view.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    and then use this form to show a pre-filtered list of report records, and 
+    optionally add some extra stuff, like a graph summarising the report 
+    records, or totals.
+    """
+
+    def _default_from_date(self):
+        max_age = datetime.now() - timedelta(days=30)
+        return fields.Datetime.to_string(max_age)
+
+    from_date = fields.Datetime('From:', default=_default_from_date)
```
Comment (nocturnalastro) [inline addons/olap/models/report_view.py]: should that be '_default_from_date'?

File: addons/olap_tutorial/models/report_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    _inherit = "olap.report.model"
+
+    _record_model = RTBRecord._name
+    _reported_model = 'stock.picking'
+
+    @api.model
+    def get_candidate_records(self):
+        """Get all pickings updated in the last 24 hours"""
+        goods_in_type = self.env.ref('stock.picking_type_in')
+        max_age = datetime.now() - timedelta(hours=24)
```
Comment (nocturnalastro) [inline addons/olap_tutorial/models/report_tutorial.py:30]: couldn't you overwrite _default_from_date and use that? That way the max_age is always coming from one place

File: addons/olap_tutorial/models/report_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+        recs = ReceiptToBin.search(self.get_report_domain())
+        self.mean_lines = self.mean(recs.mapped('num_lines'))
+        self.total_lines = sum(recs.mapped('num_lines'))
+        self.rtb_record_ids = recs
+
+    @api.multi
+    def action_compute_report(self):
+        res = super(RTBViewer, self).action_compute_report()
+        res.update({'name': _('RTB Report')})
```
Comment (nocturnalastro) [inline addons/olap_tutorial/models/report_tutorial.py:86]: would it be better to just pass the name to the super? not sure either way.

File: addons/olap/models/report_view.py (hunk 1 above)
Comment (crwr45) [reply addons/olap/models/report_view.py]: No, the default can be a variety of things, one of which is a function that will be run per-record if you want a default that is not static as we do here:
https://www.odoo.com/documentation/11.0/howtos/backend.html#default-values

File: addons/olap_tutorial/models/report_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    _reported_model = 'stock.picking'
+
+    @api.model
+    def get_candidate_records(self):
+        """Get all pickings updated in the last 24 hours"""
+        goods_in_type = self.env.ref('stock.picking_type_in')
+        max_age = datetime.now() - timedelta(hours=24)
+        res = self.env[self._reported_model].search(
+            [('picking_type_id', '=', goods_in_type.id),
+             ('write_date', '>', fields.Datetime.to_string(max_age))])
```
Comment (ws-mps) [inline addons/olap_tutorial/models/report_tutorial.py]: Use create_date instead.

**Pull request merged**